ibis/pyspark/compiler.py

```diff
@@ -0,0 +1,523 @@
+import collections
+import enum
+import functools
+
+import pyspark.sql.functions as F
+from pyspark.sql.window import Window
+
+import ibis.common.exceptions as com
+import ibis.expr.operations as ops
+import ibis.expr.types as types
+from ibis.pyspark.operations import PySparkTable
+
+
+class AggregationContext(enum.Enum):
+    ENTIRE = 0
+    WINDOW = 1
+    GROUP = 2
+
+
+class PySparkExprTranslator:
+    _registry = {}
+
+    @classmethod
+    def compiles(cls, klass):
+        def decorator(f):
+            cls._registry[klass] = f
+            return f
+
+        return decorator
+
+    def translate(self, expr, **kwargs):
+        # The operation node type the typed expression wraps
+        op = expr.op()
+
+        if type(op) in self._registry:
+            formatter = self._registry[type(op)]
+            return formatter(self, expr, **kwargs)
+        else:
+            raise com.OperationNotDefinedError(
+                'No translation rule for {}'.format(type(op))
+            )
+
+
+compiles = PySparkExprTranslator.compiles
+
+
+@compiles(PySparkTable)
+def compile_datasource(t, expr):
+    op = expr.op()
+    name, _, client = op.args
+    return client._session.table(name)
+
+
+def compile_table_and_cache(t, expr, cache):
```

Comment:
@cpcloud I found that when compiling selections I keep recompile the same TableExpr over and over again because it is referenced from multiple places. I end up adding cache for expressions that has been compiled and pass it around. 

Is this a reasonable approach. How did you deal with similar problems with Pandas backend?